### DISCUSSION TITLE
version: 5     Where did this come from?

### DISCUSSION THREAD

**Discussion by oywino** [Koenkk/zigbee2mqtt#19886] [General]
suddenly one day my Z2M crashed and would no longer start up again. The error message was:
`
MQTT retention requires protocol version 5
`
Since then I've had nothing but problems. 
Can someone explain what this `version: 5` means, and why (from where) it suddenly appeared out of the blue?
See also my post here for further details: https://github.com/Koenkk/zigbee2mqtt/discussions/19882#discussioncomment-7678374

Any help/explanations are highly appreciated.

**Comment by oywino** [comment]
Here are some more log-details about my problem - in addition to what I already wrote here: https://github.com/Koenkk/zigbee2mqtt/discussions/19882#discussioncomment-7678374

My Z2M log's last entries before the most recent crash:

```
error 2023-11-26 23:18:52: Exception while calling fromZigbee converter: Command 0x3425b4fffee6d2dc/1 manuSpecificTuya.mcuGatewayConnectionStatus({"payloadSize":1,"payload":1}, {"sendWhen":"immediate","timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":false,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false}) failed (sendZclFrameToEndpointInternal error)}
error 2023-11-27 03:49:11: Exception while calling fromZigbee converter: Command 0x3425b4fffee6d2dc/1 manuSpecificTuya.mcuGatewayConnectionStatus({"payloadSize":1,"payload":1}, {"sendWhen":"immediate","timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":false,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false}) failed (sendZclFrameToEndpointInternal error)}
error 2023-11-27 03:49:11: Exception while calling fromZigbee converter: Command 0x3425b4fffee6d2dc/1 manuSpecificTuya.mcuGatewayConnectionStatus({"payloadSize":1,"payload":1}, {"sendWhen":"immediate","timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":false,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false}) failed (sendZclFrameToEndpointInternal error)}
error 2023-11-27 03:49:11: Exception while calling fromZigbee converter: Command 0x3425b4fffee6d2dc/1 manuSpecificTuya.mcuGatewayConnectionStatus({"payloadSize":1,"payload":1}, {"sendWhen":"immediate","timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":false,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false}) failed (sendZclFrameToEndpointInternal error)}
error 2023-11-27 03:49:11: Exception while calling fromZigbee converter: Command 0x3425b4fffee6d2dc/1 manuSpecificTuya.mcuGatewayConnectionStatus({"payloadSize":1,"payload":1}, {"sendWhen":"immediate","timeout":10000,"disableResponse":false,"disableRecovery":false,"disableDefaultResponse":false,"direction":0,"srcEndpoint":null,"reservedBits":0,"manufacturerCode":null,"transactionSequenceNumber":null,"writeUndiv":false}) failed (sendZclFrameToEndpointInternal error)}
error 2023-11-27 04:04:36: Not connected to MQTT server!
error 2023-11-27 04:05:26: Not connected to MQTT server!
error 2023-11-27 04:06:21: Not connected to MQTT server!
error 2023-11-27 04:07:27: Not connected to MQTT server!
error 2023-11-27 04:08:43: Not connected to MQTT server!
error 2023-11-27 04:10:10: Not connected to MQTT server!
error 2023-11-27 04:11:51: Not connected to MQTT server!
Error: EMFILE: too many open files, open '/app/data/database.db.tmp'
     at Object.openSync (node:fs:603:3)
     at Object.writeFileSync (node:fs:2324:35)
     at Database.write (/app/node_modules/zigbee-herdsman/dist/controller/database.js:73:22)
     at Controller.databaseSave (/app/node_modules/zigbee-herdsman/dist/controller/controller.js:266:23)
     at Timeout._onTimeout (/app/node_modules/zigbee-herdsman/dist/controller/controller.js:160:57)
     at listOnTimeout (node:internal/timers:569:17)
     at processTimers (node:internal/timers:512:7)
```
There are no issues with my Mosquitto:MQTT broker. It continues to run just fine, serving other publishers.

**Comment by DubhAd** [comment]
https://www.zigbee2mqtt.io/guide/configuration/mqtt.html

```yaml
# Required: MQTT settings
mqtt:
...
  # Optional: MQTT protocol version (default: 4), set this to 5 if you
  # use the 'retention' device specific configuration
  version: 4
```

That's been there for _at least_ two years - the furthest back the current docs repo goes.

**Reply by oywino** [reply]
Hi,  and many thanks for your reply. 
I'm still very unsure about what to do. 
What does 'retain' and 'retention' actually mean in this context? 
How do I know if I need or use 'retention' specific devices? 
Why would incorrect version setting (4 vs 5) cause Z2M to refuse to start at all?

**Reply by DubhAd** [reply]
Z2M won't start because your config is invalid.

The docs have an awesome search feature that would find you details about [`retention`](https://www.zigbee2mqtt.io/guide/configuration/devices-groups.html#common-device-options) which is unrelated to `retain`

**Reply by Ra72xx** [reply]
But why does Z2M fail here instead of falling back to working settings to keep at least the users network functioning? E.g. retain could be disabled here, the logs and the web ui could contain a warning that a user's setting had to be reverted.

**Comment by oywino** [comment]
@DubhAd > Z2M won't start because your config is invalid.
> 
> The docs have an awesome search feature that would find you details about [`retention`](https://www.zigbee2mqtt.io/guide/configuration/devices-groups.html#common-device-options) which is unrelated to `retain`

Many thanks for the link.
But assuming the error message:
`
MQTT retention requires protocol version 5
`
is a result of invalid config setting for `retention` I noticed that retention was set to 120 for a new device Tuya Power Clamp TS0601 (_TZE204_cjbofhxw). I'm not sure if this was automatically set during pairing or otherwise. I certainly didn't do it.
So, I tried to delete the retention setting from this device in the Z2M UI, but that was impossible. All I could do was to set i to zero. Then i removed `version: 5` from the _configuration.yaml_ and restarted Z2M.
I had to re-pair the device in order to get it back online. 

I'll let the bridge run for a while and see if it still crashes now that the `version 5` statement is removed and no device has _retention_ set.

[Edit] Another day gone by without the `version: 5` setting, and Z2M still crashes frequently. With the `restart: unless-stopped` setting, it keeps trying forever, but every time ends up with: `Error while starting zigbee-herdsman`
Restarting the coordinator does not help.
But restarting the coordinator and issue this command from the coordinator console, helps:

`
backlog rule1 on system#boot do TCPStart 8888 endon ; rule1 1 ; template {"NAME":"Sonoff ZHABridge","GPIO":[56,208,0,209,59,58,0,0,0,0,0,0,17],"FLAG":0,"BASE":18} ; module 0
`

This helps, for a while, until it crashes again a couple of hours later. 
Why the above command helps, and why it crashes again, I don't have a clue.
Please help.
